### ISSUE TITLE
Native Infinite Sampler for Datasets

### ISSUE THREAD

**Issue by shivakanthsujit** [pytorch/pytorch#148807]
### 🚀 The feature, motivation and pitch

Currently I don't think there is a native function in PyTorch that lets you sample from a fixed length dataset through an infinite length data loader. That is, if I define my ML loop in terms of number of gradient steps that I want to do, and just keep sampling batches from the Dataloader for those many times.

From looking online this is the approach that people generally use to achieve this

```
d_iter = iter(dataloader)
for _ in range(num_updates):
  try:    
    d = next(d_iter)
  except StopIteration:
    d_iter = iter(dataloader)
    d = next(d_iter)
```
But this feels clunky and more importantly, adds overhead every time you wrap the dataloader in an iter. Depending on the dataset, the overhead of recreating the dataloader iter can be significant. In my use case it is about 25s every time and can be completely eliminated by having an infinite sampler, saving hours of training time. 

Instead I would prefer having a wrapper that wraps the original dataset which effectively allows you to sample from the dataloader an infinite number of times. How I've done this is I wrap the original dataset as part of a IterableDataset and put the sampling of indices in a while True loop. So the iter can keep yielding elements for as long as you like. I make sure to handle multiple workers and the Distributed setting. 

I have a version of an infinite sampler working and wanted to know if it could be added to PyTorch directly? I think this is a common enough use case that it warrants a PyTorch native function.

For context this is what I do
```
def __iter__(self):
         worker_info = torch.utils.data.get_worker_info()
         num_workers = 1
         worker_id = 0

         if worker_info:
             num_workers = worker_info.num_workers
             worker_id = worker_info.id

         total_processes = self.world_size * num_workers
         process_id = self.rank * num_workers + worker_id

         indices = list(range(self.dataset_len))
         if len(indices) < self.total_size:
             padding = self.total_size - len(indices)
             indices += indices[:padding]

         while True:
             if self.shuffle:
                 current_seed = self.seed + self.iteration
                 g = torch.Generator()
                 g.manual_seed(current_seed)
                 shuffle_indices = torch.randperm(len(indices), generator=g).tolist()
                 indices = [indices[i] for i in shuffle_indices]

             local_indices = indices[process_id::total_processes]
             for idx in local_indices:
                 yield self.dataset[idx]

             self.iteration += 1
```

self.iteration makes sure the ordering of the samples is properly shuffled at the end of the dataset, similar to how we use sampler.set_epoch() for the DistributedSampler. And it will be handled internally since the user wouldn't know when one pass of the dataset is complete. 

After wrapping your dataset in this iterable, you can use the DataLoader to perform batching as usual.

### Alternatives

The loop with stop iteration as mentioned above. But I feel my solution is better cause it avoids the overhead of creating the dataloader iterator multiple time.

**Comment by divyanshk** [comment]
Thanks for the issue! We can add a new sampler here (https://github.com/pytorch/pytorch/blob/main/torch/utils/data/sampler.py) which loops over infinitely. I wonder if we can construct it using a `RandomSampler` (or any other existing samplers).
